onnxruntime/core/providers/webgpu/nn/batch_norm.h

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/providers/webgpu/webgpu_kernel.h"
+#include "core/providers/webgpu/program.h"
+
+namespace onnxruntime {
+namespace webgpu {
+
+class BatchNormalizationProgram final : public Program<BatchNormalizationProgram> {
+ public:
+  BatchNormalizationProgram(float epsilon, int64_t spatial, std::string format, int64_t components) : Program{"BatchNormalization"},
+                                                                                                      epsilon_{epsilon},
+                                                                                                      spatial_{spatial},
+                                                                                                      format_{format},
+                                                                                                      components_{components} {}
+
+  Status GenerateShaderCode(ShaderHelper& sh) const override;
+
+  WEBGPU_PROGRAM_DEFINE_UNIFORM_VARIABLES({"output_size", ProgramUniformVariableDataType::Uint32});
+
+ private:
+  float epsilon_;
+  int64_t spatial_;
+  std::string format_;
+  int64_t components_;
+};
+
+class BatchNormalization final : public WebGpuKernel {
+ public:
+  BatchNormalization(const OpKernelInfo& info) : WebGpuKernel(info) {
+    epsilon_ = info.GetAttrOrDefault<float>("epsilon", 1e-5f);
+    momentum_ = info.GetAttrOrDefault<float>("momentum", 0.9f);
+    spatial_ = info.GetAttrOrDefault<int64_t>("spatial", 1);
+    training_mode_ = info.GetAttrOrDefault<int64_t>("training_mode", 0);
+    // NCHW for ai.onnx domain, NHWC for com.ms.internal.nhwc domain
+    format_ = info.GetAttrOrDefault<std::string>("format", "NHWC");
```

Comment:
Does the attribute "format" actually exist?

Reply:
Looked at this: ort can rewrite the graph to use NHWC in the kMSInternalNHWCDomain and in that case we'd use this path. It is passed in the template when we register the op.